core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveTrivialFilters.java

```diff
@@ -46,10 +47,23 @@ public Result apply(FilterNode filterNode, Captures captures, Context context)
             return Result.ofPlanNode(filterNode.getSource());
         }
 
-        if (predicate.equals(FALSE_LITERAL) || predicate instanceof NullLiteral) {
+        if (predicate.equals(FALSE_LITERAL) || isEffectivelyNull(predicate)) {
             return Result.ofPlanNode(new ValuesNode(context.getIdAllocator().getNextId(), filterNode.getOutputSymbols(), emptyList()));
         }
 
         return Result.empty();
     }
+
+    private boolean isEffectivelyNull(Expression expression)
```

Comment:
The method will return false on many expressions that are (effectively) null.

Either use constant evaluator (which i guess you don't want to use, because it's not this rule's job),
or give the method a name that "doesn't mislead". `isNull` would be better IMO.

Reply:
or maybe `isConstantNull` or `isObviouslyNull`